### DISCUSSION TITLE
bpftrace: attach kprobes related to i915 irq handlers?

### DISCUSSION THREAD

**Discussion by Binary-Eater** [bpftrace/bpftrace#2820] [Q&A]
I am trying to attach a kprobe to a `i915` irq handler which should be fine based on the kernel documentation, but I am having a hard time. From what I can tell, this should be possible. Trying to debug an issue in the irq context and figured I would learn how to use `bpftrace` to do so.

https://docs.kernel.org/trace/kprobes.html#kprobes-features-and-limitations

```sh
~ via ❄️  impure (shell) 
❯ sudo bpftrace -l '*gen8_irq*'
Place your right index finger on the fingerprint reader
hardware:*gen8_irq*:
kprobe:gen8_irq_handler
kprobe:gen8_irq_power_well_post_enable
kprobe:gen8_irq_power_well_pre_disable
software:*gen8_irq*:

~ via ❄️  impure (shell) took 7s 
❯ sudo bpftrace -l '*gen8_de*'
Place your right index finger on the fingerprint reader
hardware:*gen8_de*:
kprobe:gen8_de_irq_handler
kprobe:gen8_de_irq_postinstall
kprobe:gen8_de_pipe_underrun_mask
software:*gen8_de*:

~ via ❄️  impure (shell) took 5s 
❯ sudo bpftrace -e 'kprobe:gen8_de_irq_handler { printf("master_ctl: %x\n", arg1); }'
Place your right index finger on the fingerprint reader
Attaching 1 probe...
cannot attach kprobe, probe entry may not exist
ERROR: Error attaching probe: 'kprobe:gen8_de_irq_handler'

~ via ❄️  impure (shell) took 3s 
❯ sudo bpftrace -e 'kprobe:gen8_irq_handler { printf("test\n"); }'
Place your right index finger on the fingerprint reader
Attaching 1 probe...
cannot attach kprobe, probe entry may not exist
ERROR: Error attaching probe: 'kprobe:gen8_irq_handler'
```

```sh
~ via ❄️  impure (shell) took 3s 
❯ sudo bpftrace --info
Place your right index finger on the fingerprint reader
System
  OS: Linux 6.6.0-rc7 #1-NixOS SMP PREEMPT_DYNAMIC Tue Jan  1 00:00:00 UTC 1980
  Arch: x86_64

Build
  version: v0.18.1
  LLVM: 14.0.6
  unsafe probe: no
  bfd: yes
  libdw (DWARF support): yes

Kernel helpers
  probe_read: yes
  probe_read_str: yes
  probe_read_user: yes
  probe_read_user_str: yes
  probe_read_kernel: yes
  probe_read_kernel_str: yes
  get_current_cgroup_id: yes
  send_signal: yes
  override_return: no
  get_boot_ns: yes
  dpath: yes
  skboutput: yes

Kernel features
  Instruction limit: 1000000
  Loop support: yes
  btf: yes
  module btf: yes
  map batch: yes
  uprobe refcount (depends on Build:bcc bpf_attach_uprobe refcount): yes

Map types
  hash: yes
  percpu hash: yes
  array: yes
  percpu array: yes
  stack_trace: yes
  perf_event_array: yes

Probe types
  kprobe: yes
  tracepoint: yes
  perf_event: yes
  kfunc: yes
  iter:task: yes
  iter:task_file: yes
  iter:task_vma: yes
  kprobe_multi: no
  raw_tp_special: yes
```

**Comment by viktormalik** [comment]
Can you attach to *any* kprobe? For instance:
```
$ sudo bpftrace -e 'kprobe:vfs_read { exit(); }'
```

**Reply by Binary-Eater** [reply]
Thanks for the reply. I should have been clearer in my initial response and added an example with other kprobes. I have no issue attaching other kprobes.

```sh
~ via ❄️  impure (shell) 
❯ sudo bpftrace -e 'kprobe:vfs_read { exit(); }'
Place your right index finger on the fingerprint reader
Attaching 1 probe...
```

**Comment by danobi** [comment]
I couldn't find anything obvious in i915 subdirectories like `noinstr` or `notrace` annotations. I still kinda suspect the kernel is preventing attachment to irqs for safety reasons. Might need to printk debug the kernel to see where it's rejecting the attachment.

**Reply by Binary-Eater** [reply]
Yeah, I have come to the same conclusion. Will be able to get to this later in the week. Will post back with my findings.

**Reply by Binary-Eater** [reply]
I was able to debug the issue @danobi. The problem has to do with NixOS compiling kernels without the `CONFIG_GENERIC_IRQ_INJECTION` Kconfig option enabled by default. Once the kernel is compiled with this option, I am able to attach the related kprobes.

https://lore.kernel.org/linux-pci/20200306130623.990928309@linutronix.de/

For what it was worth, I was trying to debug the following issue using `bpftrace` which led me to this problem: https://gitlab.freedesktop.org/drm/intel/-/issues/9652

I am thinking of encouraging the NixOS folks to compile `CONFIG_GENERIC_IRQ_INJECTION` enabled by default. Arch Linux does this and either way, you need elevated privileges to perform the action of attaching the kprobe to the IRQ context. I see this option as a win for improving system debug-ability and general visibility through the use of eBPF (huge fan of Brendan Gregg's work as well as everyone else in the eBPF space, especially for debugging and profiling applications :))

Sure, using the functionality might be problematic at times but giving the user the option to inject the IRQ itself has no negative impact. Anyways, maybe this is also a good detail to add to the kernel doc page I linked.

```
+ * irq_inject_interrupt - Inject an interrupt for testing/error injection
+ * @irq:	The interrupt number
+ *
+ * This function must only be used for debug and testing purposes!
+ *
+ * Especially on x86 this can cause a premature completion of an interrupt
+ * affinity change causing the interrupt line to become stale. Very
+ * unlikely, but possible.
```

```sh
~ via ❄️  impure (shell) took 1m41s 
❯ sudo bpftrace -e 'kprobe:gen8_de_irq_handler { printf("master_ctl: %x\n", arg1); }'
Place your right index finger on the fingerprint reader
Attaching 1 probe...
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
master_ctl: 80010000
^C
```

**Reply by danobi** [reply]
Thanks for reporting back! Out of curiosity, I was trying to figure out how `CONFIG_GENERIC_IRQ_INJECTION` ties back to kprobes. Not sure I see the link. My best guess (and the only link I see) is that `CONFIG_GENERIC_IRQ_INJECTION` is depended on by `PCIEAER_INJECT` and it somehow has to do with msi-x interrupts over pcie?

**Reply by Binary-Eater** [reply]
Let me take a look and dig into the dependency to see how the `irq_inject_interrupt` is used during this kprobe attachment. Posted back after quickly debugging this without delving too deep.

**Reply by Binary-Eater** [reply]
I think your analysis on `PCIEAER_INJECT` and msi-x interrupts over pcie is likely accurate though.

https://elixir.bootlin.com/linux/v6.7-rc2/source/drivers/pci/pcie/aer_inject.c#L474

**Reply by Binary-Eater** [reply]
Ok, so to confirm because I needed to update the kernel with regards to the issue I was investigating, `CONFIG_GENERIC_IRQ_INJECTION` does not actually make a difference (which makes sense since I should have seen a kernel message related to the injection if its was used).

Basically, attaching the related kprobe is broken in commit [2ba446f82142d0d42fc5ea7bea7af581d33a7939](https://cgit.freedesktop.org/drm/drm/commit/?id=2ba446f82142d0d42fc5ea7bea7af581d33a7939) from git://anongit.freedesktop.org/drm/drm. However, it's fixed in commit [b85ea95d086471afb4ad062012a4d73cd328fa86](https://cgit.freedesktop.org/drm/drm/commit/?id=b85ea95d086471afb4ad062012a4d73cd328fa86) from git://anongit.freedesktop.org/drm/drm.

This gives me a bisection range to work with.
